### PR TITLE
dt: init at 1.2.3

### PR DESCRIPTION
**Pull request by booniepepper**
## Description of changes

An initial package for dt ([Repo](https://github.com/so-dang-cool/dt), [Website](https://dt.plumbing))

I also am the upstream owner

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by Gerg-L** [comment]
Please set `meta.mainProgram` and use `hash` instead of `sha255`

**Comment by booniepepper** [comment]
Done, thanks for the feedback @Gerg-L !

**Comment by Gerg-L** [comment]
> Done, thanks for the feedback @Gerg-L !

I will build and approve when I can get to my computer

**Review by Gerg-L — APPROVED** [review]

**Review by justinlime — APPROVED** [review]

**Comment by doronbehar** [comment]
@booniepepper now that https://github.com/NixOS/nixpkgs/pull/247437 is merged, could you remove the `zigHook = zigHook.override` line?

**Comment by booniepepper** [comment]
@doronbehar I rebased and tried just to make sure, but no, I cannot remove the `zigHook = zigHook.override` line. Just like `poop` requires it in the PR you linked (https://github.com/NixOS/nixpkgs/pull/247437), `dt` also only can build with Zig 0.11 and requires the override to build. (Unless there's some other way to configure the version of Zig)

**Comment by doronbehar** [comment]
I see. Usually, similarly to e.g `gcc9Stdenv` and other versions of stdenvs, the natural thing to do is to have a global `zig0_11Hook` that both packages use. It'd be nice to do that treewide, even in a separate PR.

Also, since there are ~20 packages using zig all togehter in Nixpkgs, I think you should make `zig_0_11` the default `zig`, and override the packages that really have no choice but to use an older zig inside them - use `zig0_10Hook` _inside_ the package, and not in the `callPackage`. This way, package maintainers will notice their package uses an old version of zig, and will hopefully ask upstream for latest zig support.

If you are willing to work on that, I'd be happy to help you.

**Comment by booniepepper** [comment]
I'm a rookie when it comes to Nix, I probably don't know enough of the basic syntax to get it done yet. @figsoda I've seen you working on Zig related commits, do you have any plans or opinions on changing the Zig hooks?

**Review by figsoda — COMMENTED** [review]
lgtm aside from the small nitpick

> I think you should make zig_0_11 the default zig, and override the packages that really have no choice but to use an older zig inside them

sounds like a good idea

> the natural thing to do is to have a global zig0_11Hook that both packages use

I'm still a bit unsure how to expose the hooks, whether it should be within the `zig` package, as a `passthru`s (`zig.hook`, `zig_0_9.setupHook`), a separate attribute (`zigHook`, `zig0_11Hook`, `zigHook0_9`), or be separated into multiple hooks that handle each phase

@AndersonTorres went with a single top-level attribute `zigHook` + overrides in the PR that introduced it, so I would like to hear from them as well

**Comment by doronbehar** [comment]
> I'm still a bit unsure how to expose the hooks, whether it should be within the `zig` package, as a `passthru`s (`zig.hook`, `zig_0_9.setupHook`), a separate attribute (`zigHook`, `zig0_11Hook`, `zigHook0_9`), or be separated into multiple hooks that handle each phase

There's also the option of using `zig.setupHook = ./setup-hook.sh`, without a passthru. This is the most common option in nixpkgs of them all, just like simply adding `cmake` to `nativeBuildInputs` automagically makes `mkDerivation` use `cmake.setupHook`'s phases to build the package.

**Comment by figsoda** [comment]
> There's also the option of using zig.setupHook = ./setup-hook.sh, without a passthru.

This is what I meant by "within the `zig` package"

**Comment by doronbehar** [comment]
> > There's also the option of using zig.setupHook = ./setup-hook.sh, without a passthru.
> 
> This is what I meant by "within the `zig` package"

That's what I assumed, but I think that defining it in the `passthru` will not produce the behavior we depict, but I'm not sure.

**Comment by figsoda** [comment]
ah that was supposed to be a different option

to clarify, here are the options:
- within the `zig` package
- as `passthru`s (`zig.hook`, `zig_0_9.setupHook`)
- as separate attributes (`zigHook`, `zig0_11Hook`, `zigHook0_9`)
- be separated into multiple hooks that handle each phase

**Review by figsoda — APPROVED** [review]
lgtm for a squash merge

**Review by Gerg-L — APPROVED** [review]
LGTM

**Comment by booniepepper** [comment]
Could I help capture any of the conversation above on how to package Zig somewhere else...? Some of it's going over my head right now, but should it be in an issue or a wiki or something?

**Comment by AndersonTorres** [comment]
> There's also the option of using zig.setupHook = ./setup-hook.sh, without a passthru. 

And a useless (and possibly mass) rebuild as a gift. Nope, I'll pass.

File: pkgs/tools/text/dt/default.nix
Comment (AndersonTorres) [inline pkgs/tools/text/dt/default.nix:25]: Please, can you add a longDescription? I copy-pasted this from their website:

```text
dt is a utility and programming language. The utility is intended for ergonomic in-the-shell execution. The language is straightforward (in the most literal sense) with a minimal syntax that allows for high-level, higher-order programming.

It's meant to supplement (not replace!) other tools like awk, sed, xargs, and shell built-ins. Something like the Perl one-liners popular yesteryear, but hopefully easier to read and reason through.

In particular dt is intended to be generally useful, with zero pretense of elegance. 
```

(obviously, break each paragraph in the 80-column limit.)

**Review by AndersonTorres — CHANGES_REQUESTED** [review]
Ping me 02 hours to merge

**Comment by AndersonTorres** [comment]
Don't forget to squash the commits into one.

File: pkgs/tools/text/dt/default.nix
Comment (booniepepper) [reply pkgs/tools/text/dt/default.nix]: Thanks for looking, I added it

**Review by AndersonTorres — APPROVED** [review]

**Comment by figsoda** [comment]
> I think you should make `zig_0_11` the default `zig`, and override the packages that really have no choice but to use an older zig inside them

#248153

**Pull request merged**
